File: src/main/webapp/js/systems/los.js
import { TILE } from '../core/constants.js';
import { isSolidChar, solidAt } from './physics.js';

export function hasLineOfSight(state, x0,y0,x1,y1){
  let cx=Math.floor(x0/TILE), cy=Math.floor(y0/TILE);
  const tx=Math.floor(x1/TILE), ty=Math.floor(y1/TILE);
  const dx=Math.sign(tx-cx), dy=Math.sign(ty-cy);
  const nx=Math.abs(tx-cx), ny=Math.abs(ty-cy);
  let err=nx-ny, guard=0;
  while(!(cx===tx && cy===ty) && guard++<1200){
    if(isSolidChar(solidAt(state,cx,cy))) return false;
    const e2=2*err; if(e2>-ny){ err-=ny; cx+=dx; } if(e2< nx){ err+=nx; cy+=dy; }
  }
  return true;
}

File: core/src/main/kotlin/io/github/panda17tk/arpg/pathfinding/Los.kt
package io.github.panda17tk.arpg.pathfinding

import io.github.panda17tk.arpg.map.TileMap
import io.github.panda17tk.arpg.sim.Tuning
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.sign

/** Integer-grid line of sight (Bresenham). False if any solid tile lies on the line. */
object Los {
    fun hasLineOfSight(map: TileMap, x0: Float, y0: Float, x1: Float, y1: Float): Boolean {
        var cx = floor(x0 / Tuning.TILE).toInt()
        var cy = floor(y0 / Tuning.TILE).toInt()
        val tx = floor(x1 / Tuning.TILE).toInt()
        val ty = floor(y1 / Tuning.TILE).toInt()
        val dx = sign((tx - cx).toFloat()).toInt()
        val dy = sign((ty - cy).toFloat()).toInt()
        val nx = abs(tx - cx)
        val ny = abs(ty - cy)
        var err = nx - ny
        var guard = 0
        while (!(cx == tx && cy == ty) && guard++ < 1200) {
            if (map.solidAt(cx, cy)) return false
            val e2 = 2 * err
            if (e2 > -ny) { err -= ny; cx += dx }
            if (e2 < nx) { err += nx; cy += dy }
        }
        return true
    }
}
